### PR TITLE
fix: auto mode and listmodels now work correctly with custom provider

### LINKED ISSUE #327: Local models not detected properly
### Project Version

9.1.3

### Bug Description

	"CUSTOM_API_URL": "http://10.75.54.131:8080/v1",
	"CUSTOM_MODEL_NAME": "model",

I have this configuration, which if you check v1/models, can see hosts one model named "model".

{"object":"list","data":[{"id":"model","object":"model","created":1763064981,"owned_by":"sglang","root":"model","parent":null,"max_model_len":131072}]}

When I ask Zen what models are available, it thinks I have 3 Llama models, which makes me feel like it is reverting to default setting.

Image

It also doesn't seem to use local/custom model at all.  If I use consensus, it will ask gemini flash and pro, but ignore local/custom.  If I use "chat" it will say cli isn't installed and won't use any of them.

"models":[{"model":"gemini-2.5-pro","stance":"neutral"},{"model":"gemini-2.0-flash","stance":"neutral"}],"use_assistant_model":true}}

I was not using openai or open router, I did add openai when creating this because of the sanity check boxes, but I did not add openrouter but I had to check it anyway.

### Relevant Log Output

```shell

```

### Operating System

Linux

### Sanity Checks

- [x] I have searched the existing issues and this is not a duplicate.
- [x] I am using `GEMINI_API_KEY`
- [x] I am using `OPENAI_API_KEY`
- [x] I am using `OPENROUTER_API_KEY`
- [x] I am using `CUSTOM_API_URL`

### LINKED ISSUE #344: Failed to connect to zen mcp server when default_model is set to "auto"
### Project Version

9.4.1

### Bug Description

Claude-code can't connect when using a custom endpoint:

> 3. zen                            ✘ failed · Enter to view details

I was using the "auto" model. But apparently that does not work when using a custom endpoint.
Maybe there could be a better error message when this happens?

### Relevant Log Output

```shell
2025-11-23 17:34:27,432 - root - INFO - Process PID: 530568
2025-11-23 17:34:27,433 - __main__ - INFO - All tools enabled (DISABLED_TOOLS not set)
2025-11-23 17:34:27,435 - __main__ - INFO - Custom API endpoint found: https://custom.domain/v1 with model llama3.2
2025-11-23 17:34:27,435 - __main__ - INFO - Registered providers: custom
2025-11-23 17:34:27,435 - __main__ - INFO - Available providers: Custom API (https://custom.domain/v1)
2025-11-23 17:34:27,435 - __main__ - INFO - No model restrictions configured - all models allowed
2025-11-23 17:34:27,435 - root - INFO - Initializing Custom provider with endpoint: https://custom.domain/v1
2025-11-23 17:34:27,435 - root - INFO - Model allow-list not configured for Custom API - all models permitted. To restrict access, set CUSTOM_ALLOWED_MODELS with comma-separated model names.
2025-11-23 17:34:27,435 - root - INFO - Using extended timeouts for custom endpoint: https://custom.domain/v1
```

### Operating System

Linux

### Sanity Checks

- [x] I have searched the existing issues and this is not a duplicate.
- [x] I am using `GEMINI_API_KEY`
- [x] I am using `OPENAI_API_KEY`
- [x] I am using `OPENROUTER_API_KEY`
- [x] I am using `CUSTOM_API_URL`

### PR DESCRIPTION
**Pull request by DragonFSKY**
## Summary

- Make `CUSTOM_MODEL_NAME` visible to every custom model registry, including `listmodels`.
- Prefer that configured model when auto mode selects from the custom provider.

Fixes #327.
Fixes #344.

## Testing

- `python -m pytest tests/test_auto_mode_custom_provider_only.py -q` — 4 passed
- Ruff, Black, and isort checks passed for the changed files.

### CODE REVIEW

**Comment by DragonFSKY** [comment]
Hi @gemini-code-assist, I've addressed your code review feedback:

### Changes Made:

1. **Redundant expression** ✅ Fixed
   - Removed `or ""` and added `.strip()` to handle whitespace

2. **Magic numbers** ✅ Addressed
   - Changed `intelligence_score` from 10 to 6 to match `llama3.2` in `custom_models.json`
   - Added comment explaining the choice

3. **Encapsulation** ✅ Already resolved
   - The injection logic is now inside `CustomEndpointModelRegistry` class, operating on its own attributes

4. **Test assertion** ✅ Improved
   - Changed from weak assertion (`!= "gemini-2.5-flash"`) to strong assertion (`== unique_model`)
   - Added new test `test_get_preferred_model_prioritizes_custom_model_name` to verify alphabetical ordering is overridden

### Additional Improvements:

- Added `get_preferred_model()` to `CustomProvider` to ensure user-specified `CUSTOM_MODEL_NAME` is prioritized over alphabetical ordering
- Added warning log when `CUSTOM_MODEL_NAME` is configured but not found in allowed models
- Added `.strip()` to handle potential whitespace in environment variable

/gemini review

**Comment by DragonFSKY** [comment]
Addressed the remaining suggestions:

1. **Magic numbers** ✅ Extracted to module-level constants:
   - `_DEFAULT_CONTEXT_WINDOW = 128_000`
   - `_DEFAULT_MAX_OUTPUT_TOKENS = 16_000`
   - `_DEFAULT_INTELLIGENCE_SCORE = 6`

2. **Test code duplication** ✅ Extracted to helper:
   - Added `_OTHER_PROVIDER_KEYS` constant
   - Added `_clear_other_provider_keys()` helper function
   - Replaced 6 occurrences of duplicated code

/gemini review

**Comment by DragonFSKY** [comment]
Addressed the `reset_registry()` suggestion:

- Replaced `CustomProvider._registry = None` with `CustomProvider.reset_registry()` in 3 places
- Keeps `or ""` as defensive programming (consistent with other files in the project)

/gemini review

File: providers/custom.py [hunk 1]
```diff
@@ -81,11 +81,56 @@ def __init__(self, api_key: str = "", base_url: str = "", **kwargs):
...
+    # Dynamic model injection
+    # ------------------------------------------------------------------
+    def _inject_custom_model_if_needed(self) -> None:
+        """Inject CUSTOM_MODEL_NAME into registry if not already present.
+
+        This ensures auto mode works when only CUSTOM_API_URL is configured,
+        even if the model isn't declared in custom_models.json.
+        See: https://github.com/BeehiveInnovations/zen-mcp-server/issues/344
+        """
+        custom_model_name = get_env("CUSTOM_MODEL_NAME", "") or ""
```
[the remark this answers is not in the document]
Comment (DragonFSKY) [reply providers/custom.py]: Thanks for the cleanup suggestion. The redundant fallback is removed in f626a07; the explicit default from get_env is used directly.

File: providers/custom.py [hunk 2]
```diff
@@ -81,11 +81,56 @@ def __init__(self, api_key: str = "", base_url: str = "", **kwargs):
...
+            model_name=custom_model_name,
+            friendly_name=f"Custom ({custom_model_name})",
+            description="Custom model via CUSTOM_MODEL_NAME environment variable",
+            context_window=128000,  # Conservative default
+            max_output_tokens=16000,  # Conservative default
+            intelligence_score=10,  # Mid-range default
+            supports_extended_thinking=False,
+            supports_json_mode=False,
+            supports_function_calling=False,
+            supports_images=False,
```
[the remark this answers is not in the document]
Comment (DragonFSKY) [reply providers/custom.py]: Thanks. This was addressed by defining the injected capability defaults as descriptive module-level constants.

File: providers/custom.py [hunk 3]
```diff
@@ -81,11 +81,56 @@ def __init__(self, api_key: str = "", base_url: str = "", **kwargs):
...
+        )
+
+        # Inject into registry maps
+        self._registry.model_map[custom_model_name] = capability
+        self._registry.alias_map[custom_model_name.lower()] = custom_model_name
```
[the remark this answers is not in the document]
Comment (DragonFSKY) [reply providers/custom.py]: Thanks for flagging the coupling. The injection was moved into CustomEndpointModelRegistry itself, so CustomProvider no longer reaches into registry maps.

File: tests/test_auto_mode_custom_provider_only.py [hunk 4]
```diff
@@ -206,3 +206,56 @@ def test_auto_mode_fallback_with_custom_only_should_work(self):
...
+            assert resolved.model_name == unique_model
+
+            # Check auto mode returns the injected model
+            from tools.models import ToolModelCategory
+
+            fallback_model = ModelProviderRegistry.get_preferred_fallback_model(ToolModelCategory.FAST_RESPONSE)
+            print(f"Fallback model with injected CUSTOM_MODEL_NAME: {fallback_model}")
+
+            # Should return either the injected model or an existing model, not gemini-2.5-flash
+            assert fallback_model != "gemini-2.5-flash", "Should not fall back to Gemini when custom model is injected"
```
[the remark this answers is not in the document]
Comment (DragonFSKY) [reply tests/test_auto_mode_custom_provider_only.py]: Thanks. The regression test now asserts the exact injected model is selected, rather than only excluding the Gemini fallback.

File: tests/test_auto_mode_custom_provider_only.py [hunk 5]
```diff
@@ -206,3 +211,205 @@ def test_auto_mode_fallback_with_custom_only_should_work(self):
...
+        test_env = {
+            "CUSTOM_API_URL": "http://localhost:11434/v1",
+            "CUSTOM_API_KEY": "",
+            # CUSTOM_MODEL_NAME intentionally not set
+        }
+
+        with patch.dict(os.environ, test_env, clear=False):
+            # Clear CUSTOM_MODEL_NAME if set
+            if "CUSTOM_MODEL_NAME" in os.environ:
+                del os.environ["CUSTOM_MODEL_NAME"]
```
[the remark this answers is not in the document]
Comment (DragonFSKY) [reply tests/test_auto_mode_custom_provider_only.py]: Thanks for identifying the isolation gap. f626a07 centrally saves, clears, and restores CUSTOM_MODEL_NAME for every test in this class.

File: providers/custom.py [hunk 6]
```diff
@@ -116,6 +133,40 @@ def get_provider_type(self) -> ProviderType:
...
+        regardless of the task type. They may only have this one model available.
+
+        Args:
+            category: The tool category requiring a model (ignored for custom provider)
+            allowed_models: Pre-filtered list of models allowed by restrictions
+
+        Returns:
+            CUSTOM_MODEL_NAME if configured and available, None otherwise
+        """
+        custom_model = (get_env("CUSTOM_MODEL_NAME", "") or "").strip()
```
[the remark this answers is not in the document]
Comment (DragonFSKY) [reply providers/custom.py]: Thanks. The redundant fallback is removed in f626a07.

File: providers/registries/custom.py [hunk 7]
```diff
@@ -17,6 +26,56 @@ def __init__(self, config_path: str | None = None) -> None:
...
+        super().reload()
+        self._inject_custom_model_if_needed()
+
+    def _inject_custom_model_if_needed(self) -> None:
+        """Inject CUSTOM_MODEL_NAME into registry if not already present.
+
+        This ensures listmodels and other tools can see user-defined models
+        even if they're not declared in custom_models.json.
+        """
+        custom_model_name = (get_env("CUSTOM_MODEL_NAME", "") or "").strip()
```
[the remark this answers is not in the document]
Comment (DragonFSKY) [reply providers/registries/custom.py]: Thanks. The redundant fallback is removed in f626a07.

File: providers/registries/custom.py [hunk 8]
```diff
@@ -17,6 +26,56 @@ def __init__(self, config_path: str | None = None) -> None:
...
+            friendly_name=f"Custom ({custom_model_name})",
+            description="Custom model via CUSTOM_MODEL_NAME environment variable",
+            context_window=128000,  # Conservative default
+            max_output_tokens=16000,  # Conservative default
+            intelligence_score=6,  # Conservative default, matches llama3.2 in custom_models.json
+            supports_extended_thinking=False,
+            supports_json_mode=False,
+            supports_function_calling=False,
+            supports_images=False,
+        )
```
[the remark this answers is not in the document]
Comment (DragonFSKY) [reply providers/registries/custom.py:72]: Thanks for the idea. I kept the injected entry intentionally conservative and the environment surface minimal; users needing precise capabilities can define the model in custom_models.json.

[the remark this answers is not in the document]
Comment (DragonFSKY) [reply providers/registries/custom.py:72]: Thanks. These values are now named module-level constants, including a note explaining the intelligence-score baseline.

File: tests/test_auto_mode_custom_provider_only.py [hunk 9]
```diff
@@ -206,3 +211,205 @@ def test_auto_mode_fallback_with_custom_only_should_work(self):
...
+            "CUSTOM_API_KEY": "",
+            "CUSTOM_MODEL_NAME": unique_model,
+            "DEFAULT_MODEL": "auto",
+        }
+
+        with patch.dict(os.environ, test_env, clear=False):
+            # Clear other provider keys
+            for key in ["GEMINI_API_KEY", "OPENAI_API_KEY", "XAI_API_KEY", "OPENROUTER_API_KEY", "DIAL_API_KEY"]:
+                if key in os.environ:
+                    del os.environ[key]
```
[the remark this answers is not in the document]
Comment (DragonFSKY) [reply tests/test_auto_mode_custom_provider_only.py]: Thanks. The provider-key cleanup is now centralized in _OTHER_PROVIDER_KEYS and _clear_other_provider_keys.

File: providers/custom.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (DragonFSKY) [reply providers/custom.py]: Thanks. This duplicate fallback is removed in f626a07.

File: providers/registries/custom.py [hunk 10]
```diff
@@ -17,6 +32,56 @@ def __init__(self, config_path: str | None = None) -> None:
...
+        super().reload()
+        self._inject_custom_model_if_needed()
+
+    def _inject_custom_model_if_needed(self) -> None:
+        """Inject CUSTOM_MODEL_NAME into registry if not already present.
+
+        This ensures listmodels and other tools can see user-defined models
+        even if they're not declared in custom_models.json.
+        """
+        custom_model_name = (get_env("CUSTOM_MODEL_NAME", "") or "").strip()
```
[the remark this answers is not in the document]
Comment (DragonFSKY) [reply providers/registries/custom.py]: Thanks. This duplicate fallback is removed in f626a07.

File: tests/test_auto_mode_custom_provider_only.py [hunk 11]
```diff
@@ -206,3 +218,197 @@ def test_auto_mode_fallback_with_custom_only_should_work(self):
...
+
+            # Reload config
+            import config
+
+            importlib.reload(config)
+
+            # Reset registry to force re-initialization
+            from providers.custom import CustomProvider
+
+            CustomProvider._registry = None
```
[the remark this answers is not in the document]
Comment (DragonFSKY) [reply tests/test_auto_mode_custom_provider_only.py]: Thanks. The test now uses the public CustomProvider.reset_registry() helper.

File: tests/test_auto_mode_custom_provider_only.py [hunk 12]
```diff
@@ -206,3 +218,197 @@ def test_auto_mode_fallback_with_custom_only_should_work(self):
...
+        }
+
+        with patch.dict(os.environ, test_env, clear=False):
+            # Clear other provider keys
+            _clear_other_provider_keys()
+
+            # Reset registry
+            from providers.custom import CustomProvider
+
+            CustomProvider._registry = None
```
[the remark this answers is not in the document]
Comment (DragonFSKY) [reply tests/test_auto_mode_custom_provider_only.py]: Thanks. The test now uses the public CustomProvider.reset_registry() helper.

File: tests/test_auto_mode_custom_provider_only.py [hunk 13]
```diff
@@ -206,3 +218,197 @@ def test_auto_mode_fallback_with_custom_only_should_work(self):
...
+            if "CUSTOM_MODEL_NAME" in os.environ:
+                del os.environ["CUSTOM_MODEL_NAME"]
+
+            # Clear other provider keys
+            _clear_other_provider_keys()
+
+            # Reset registry
+            from providers.custom import CustomProvider
+
+            CustomProvider._registry = None
```
[the remark this answers is not in the document]
Comment (DragonFSKY) [reply tests/test_auto_mode_custom_provider_only.py]: Thanks. The test now uses the public CustomProvider.reset_registry() helper.

File: providers/custom.py [hunk 14]
```diff
@@ -116,6 +133,40 @@ def get_provider_type(self) -> ProviderType:
...
+        if not custom_model:
+            return None
+
+        if custom_model in allowed_models:
+            logging.debug(f"Custom provider preferring user-specified model: {custom_model}")
+            return custom_model
```
[the remark this answers is not in the document]
Comment (DragonFSKY) [reply providers/custom.py]: Thanks for catching this edge case. f626a07 now matches CUSTOM_MODEL_NAME case-insensitively, returns the registry canonical spelling, and includes a regression test.

File: tests/test_auto_mode_custom_provider_only.py [hunk 15]
```diff
@@ -206,3 +218,197 @@ def test_auto_mode_fallback_with_custom_only_should_work(self):
...
+        test_env = {
+            "CUSTOM_API_URL": "http://localhost:11434/v1",
+            "CUSTOM_API_KEY": "",
+            # CUSTOM_MODEL_NAME intentionally not set
+        }
+
+        with patch.dict(os.environ, test_env, clear=False):
+            # Clear CUSTOM_MODEL_NAME if set
+            if "CUSTOM_MODEL_NAME" in os.environ:
+                del os.environ["CUSTOM_MODEL_NAME"]
```
[the remark this answers is not in the document]
Comment (DragonFSKY) [reply tests/test_auto_mode_custom_provider_only.py]: Thanks. f626a07 moves CUSTOM_MODEL_NAME into the shared per-test environment lifecycle, so manual cleanup is no longer needed.

**Comment by DragonFSKY** [comment]
Maintenance update: I reviewed every outstanding thread and refreshed both the implementation and PR description. The final edge-case fixes make CUSTOM_MODEL_NAME selection case-insensitive while returning the canonical registry spelling, centralize test environment cleanup, and avoid duplicate registry injection during initialization.

Local verification: the targeted suite passed 9/9; the full Python 3.11 non-integration suite passed with 856 passed, 4 skipped, and 16 deselected (19 subtests). Black checked 247 files and Ruff passed. All 17 review threads are now answered and resolved, and the latest GitHub Actions run is fully green.

Thanks for the detailed review suggestions.
